### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by yoyomolinas**
**Changes:**
- Setup go module
- Add payment token data types
- Script to generate link

Any suggestions as to where to add further comments are welcome.

### CODE REVIEW

**Review by ckettell — APPROVED** [review]
1️⃣

**Comment by yoyomolinas** [comment]
@oliver-hohn Merging seems to be restricted to authorized users only. Any further setup required?

File: go/payment_token.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	PayeeAccountNumber string
+}
+
+// Generates a valid payment token with given details. Returns a signed JWT.
+func GeneratePaymentTokenFromDetails(paymentDetails PaymentDetails) string {
+	// JTI should be a Version 4 UUID
+	jti, err := uuid.NewRandom()
+	if err != nil {
+		panic(err)
+	}
```
Comment (oliver-hohn) [inline go/payment_token.go:65]: I think you can use `uuid.NewString()`. That should return a UUID or panic, ref: https://pkg.go.dev/github.com/google/uuid#NewString.

**Review by oliver-hohn — APPROVED** [review]
I think we should also include a `README` for how to run this command.

**Comment by oliver-hohn** [comment]
> @oliver-hohn Merging seems to be restricted to authorized users only. Any further setup required?

CC: @tobypinder @glasseyes42

**Comment by glasseyes42** [comment]
Add a codeowners?

**Comment by glasseyes42** [comment]
Though that may not help with **this** PR ... and whilst I'm unsure if an admin user can override anything - I'm not an admin 🤷  so I can't even check ... @tobypinder ?

**Comment by tobypinder** [comment]
> > @oliver-hohn Merging seems to be restricted to authorized users only. Any further setup required?
> 
> CC: @tobypinder @glasseyes42 Is anything missing in: https://github.com/creditkudos/repos/blob/master/repos/payments-auth-sample-implementations.tf?

No and the last deploy was successful - https://app.circleci.com/pipelines/github/creditkudos/repos/308/workflows/7fef4c53-360b-44c2-9123-06fa70afa16f

Bearing in mind that the project is public (and thus cannot use CI) everything is in place - are you sure this is something blocking you from merging this PR because of user permissions rather than the standard restrictions of not being capable of merging `master` directly? Github sometimes gets confused about _why_ something cannot be merged in the UI.

**Comment by tobypinder** [comment]
Criteria for being able to merge into master should be: 
* [X] At least one approved review 
* [X} At least one approved review for all codeowners (no Codeowners are yet set so this should be met)
* [X] Passed all required status check (CI) steps ([no CI requirements](https://github.com/creditkudos/repos/blob/master/repos/payments-auth-sample-implementations.tf#L16) are or should be set for this project so this should be met)
If this is still preventing merge can you attach a screenshot of the  "Changes Approved" section showing the button?

File: go/cmd/generate_payment_link.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package main
```
Comment (hatstand) [inline go/cmd/generate_payment_link.go:1]: This file should be in a further subdirectory named after it.

File: go/cmd/generate_payment_link.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	. "github.com/creditkudos/payments-auth-sample-implementations/go"
```
Comment (hatstand) [inline go/cmd/generate_payment_link.go:8]: I'd avoid importing the entire package like this as it makes the code less readable.

File: go/cmd/generate_payment_link.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		PayeeName:          "Creditor Name",
+		PayeeSortCode:      "000000",
+		PayeeAccountNumber: "12345678",
+	}
+	paymentToken := GeneratePaymentTokenFromDetails(paymentDetails)
+
+	baseUri := "http://payments-auth.creditkudos.com/oauth/authorize"
+
+	// Generate the url to enter the PIS journey. 3 query params required: clientId, redirectUri, and the payment token.
+	u := fmt.Sprintf("%s?client_id=%s&redirect_uri=%s&payment_token=%s", baseUri, c.ClientID, c.RedirectURI, paymentToken)
```
Comment (hatstand) [inline go/cmd/generate_payment_link.go:49]: Use go's url support.

File: go/payment_token.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+type PaymentTokenClaims struct {
+	jwt.StandardClaims
+	CustomerEmail      string              `json:"customerEmail"`
+	PaymentType        PaymentType         `json:"paymentType"`
+	SFDDPaymentDetails PaymentDetailsClaim `json:"singleFutureDatedDomesticPaymentDetails"`
+}
+
+// Details required to generate a valid payment token.
+// Used to structure GeneratePaymentTokenFromDetails parameters.
+type PaymentDetails struct {
```
Comment (hatstand) [inline go/payment_token.go:45]: These should all have `json` tags for clarity.

File: go/payment_token.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	ClientId           string
```
Comment (hatstand) [inline go/payment_token.go:46]: `ClientID`

**Comment by yoyomolinas** [comment]
@tobypinder 
Screenshot 2021-04-09 at 17 09 41

File: go/payment_token.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+				AccountNumber: paymentDetails.PayeeAccountNumber,
+			},
+		}
+	default:
+		panic(fmt.Errorf("Unkown payment type %s", paymentDetails.PaymentType))
+	}
+
+	// Generate and sign JWT
+	signingKey := []byte(paymentDetails.ClientSecret)
+	paymentToken, _ := jwt.NewWithClaims(jwt.SigningMethodHS256, claims).SignedString(signingKey)
```
Comment (hatstand) [inline go/payment_token.go:107]: This error check isn't one we should skip.

File: go/payment_token.go (hunk 5 above)
Comment (yoyomolinas) [reply go/payment_token.go:45]: I thought JSON tags were necessary for types we intend to encode in JSON format. `PaymentDetails` only enables prettier parameter passing to the `GeneratePaymentTokenFromDetails` function. Should it still have JSON tags?

Comment (hatstand) [reply go/payment_token.go:45]: My bad, it's fine then.

**Comment by tobypinder** [comment]
Merge issue should be fixed - is related to upstream issues with Github Terraform Provider for repo provisioning.

**Pull request merged**